### PR TITLE
TensorIterator norm update

### PR DESCRIPTION
**Pull request by jjsjann123**
special case for norm out where p == 2. Instead of calling `pow`,
we use multiplication as a faster code path.

### CODE REVIEW

**Comment by jjsjann123** [comment]
@ngimel 
Sorry for the delay, had a quite fun vacation :smile: 

Here's the perf measured with your script:
```
import torch
import time
nlayers = 10
params = []
size = 1024
for _ in range(nlayers):
    params.append(torch.randn(size, device="cuda", requires_grad=True))
    params.append(torch.randn((size, size), device="cuda", requires_grad=True))
def bench(fn, nrep, a):
    #dry tun
    c=fn(a)
    torch.cuda.synchronize()
    start = time.time()
    for _ in range(nrep):
        c=fn(a)
    torch.cuda.synchronize()
    end = time.time()
    return((end-start)/nrep)
def norm_orig(parameters):
    total_norm = 0.0
    norm_type = 2
    for p in parameters:
        param_norm = p.data.norm(2)
        total_norm += param_norm ** norm_type
    return total_norm
def norm_1(parameters):
    norm_type = 2
    total_norm = torch.sum(torch.stack([torch.norm(p.data,norm_type) for p in parameters]) ** norm_type)
    return total_norm
nrep = 100
t0 = bench(norm_orig, nrep, params)
t1 = bench(norm_1, nrep, params)
#print(t0, t1)
size = 1024
for _ in range(20):
    nrep = 1000
    a=torch.randn(size, device="cuda")
    t = bench(torch.norm, nrep, a)
    print(size, t, size*a.element_size()*1e-9/t)
    size *= 2
```
before:
```
1024 1.1724233627319335e-05 0.34936185427554656
2048 1.1457204818725586e-05 0.7150086019769015
4096 1.1605262756347657e-05 1.4117732914783467
8192 1.1620998382568359e-05 2.8197233078659067
16384 1.8230676651000977e-05 3.59481994303276
32768 3.218817710876465e-05 4.072054144511026
65536 5.995512008666992e-05 4.372337168552909
131072 1.5023469924926757e-05 34.897929880373894
262144 1.4912128448486328e-05 70.31699087238194
524288 1.5024662017822266e-05 139.58064397803804
1048576 2.115511894226074e-05 198.26425989131198
2097152 3.5564422607421877e-05 235.8707771695806
4194304 5.4831266403198245e-05 305.97899885495633
8388608 9.476947784423829e-05 354.0637002891329
16777216 0.00017866992950439452 375.6024541239292
33554432 0.0003462977409362793 387.57898806130765
67108864 0.0006764407157897949 396.8351545583439
134217728 0.0013396568298339843 400.7525659138626
268435456 0.0026634774208068846 403.13532061958176
536870912 0.005314234972000122 404.1002438384373
```

vs after
```
1024 1.1579036712646485e-05 0.3537427250339744
2048 1.1408329010009766e-05 0.7180718572204807
4096 1.1632680892944335e-05 1.4084457530282226
8192 1.1588335037231445e-05 2.8276710929328264
16384 1.1737823486328125e-05 5.583317901852454
32768 1.210165023803711e-05 10.830919537570432
65536 1.2956142425537109e-05 20.23318294829046
131072 1.5238046646118163e-05 34.40650971714675
262144 1.5089035034179687e-05 69.4925817074959
524288 1.4934301376342774e-05 140.42518274889446
1048576 1.497197151184082e-05 280.14373368816985
2097152 1.795697212219238e-05 467.1504718567124
4194304 2.9098272323608398e-05 576.5708635006514
8388608 4.897642135620117e-05 685.1140012039996
16777216 8.812785148620605e-05 761.4943842186374
33554432 0.00016767740249633788 800.4520943299522
67108864 0.00032480716705322264 826.4456059740037
134217728 0.0006382706165313721 841.1336791870191
268435456 0.0012749519348144531 842.182198936201
536870912 0.0025335981845855714 847.6022997905924
```

**Comment by ngimel** [comment]
What about norm ops in THC https://github.com/pytorch/pytorch/blob/68f378210666c079326139b6557e8a601fc89ebf/aten/src/THC/generic/THCTensorMathReduce.cu#L180-L279 - are those still used, or can they be removed?

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit 782b4707:


**None of the build failures appear to be your fault**.

* **1/2** [broken upstream](https://dr.pytorch.org/master-timeline.html?count=1&sha1=492ca46e71b9f5071a090bc78601c6621f400c9c&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35) at merge base 492ca46e since Jan 06
  <details>
  <summary>
  You may want to rebase on the <code>viable/strict</code> branch (expand for instructions)
  </summary>
  
  If your commit is newer than `viable/strict`, you can try basing on an older, stable commit:
  ```
  git fetch origin viable/strict
  git rebase --onto viable/strict $(git merge-base origin/master HEAD)
  ```
  If your commit is older than `viable/strict`:
  ```
  git fetch origin viable/strict
  git rebase viable/strict
  ```
  Check out the [recency history](https://dr.pytorch.org/master-viable-commits.html) of this "viable master" tracking branch.
  
  </details>
  
* **1/2** recognized as flaky :snowflake:
    * Re-run these jobs?


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=782b470773e3ec480f5257a031f6ce97e021a0a9).

### :snowflake: 1 failure recognized as flaky
The following build failures have been detected as flaky and **may not be your fault**:
#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4147913) pytorch_windows_test1 (1/1)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=80981052) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=80981052)) :snowflake:
<details>
<summary>
<code>AssertionError: tensor(0.0062, device='cuda:0', grad_fn=<MaxBackward1>) not less than or equal to 1e-05 :</code>
</summary>

```
---------------------------------------------------------------------- 
Traceback (most recent call last): 
  File "test_nn.py", line 5699, in test_RNN_cudnn_weight_norm 
    self.assertEqual(m(input), expected_output) 
  File "C:\Users\circleci\project\test\common_utils.py", line 867, in assertEqual 
    allow_inf=allow_inf) 
  File "C:\Users\circleci\project\test\common_utils.py", line 846, in assertEqual 
    assertTensorsEqual(x, y) 
  File "C:\Users\circleci\project\test\common_utils.py", line 816, in assertTensorsEqual 
    self.assertLessEqual(max_err, prec, message) 
AssertionError: tensor(0.0062, device='cuda:0', grad_fn=<MaxBackward1>) not less than or equal to 1e-05 :  
 
====================================================================== 
FAIL: test_embedding_max_norm_device_cuda_float16 (__main__.TestNNDeviceTypeCUDA) 
---------------------------------------------------------------------- 
Traceback (most recent call last): 
  File "C:\Users\circleci\project\test\common_device_type.py", line 197, in instantiated_test 
    result = test(self, device_arg, dtype) 
  File "C:\Users\circleci\project\test\common_device_type.py", line 389, in only_fn 
    return fn(slf, device, *args, **kwargs) 
  File "test_nn.py", line 9585, in test_embedding_max_norm_device 
```

</details>

### :construction: 1 upstream failure recognized by patterns:
These builds matched patterns, but were probably **caused by upstream breakages**:
* [pytorch_windows_test2](https://dr.pytorch.org/build-details.html?build_id=80981423) from Jan 06 until Jan 06



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

**Comment by jjsjann123** [comment]
I don't see any use of `THCTensor_norm` / `THCTensor_normall`.
Removing the code seems to compile & run fine.

**Comment by ngimel** [comment]
What about their TH counterparts? CPU reductions go to TensorIterator too, right?

**Comment by ngimel** [comment]
Also, looks like cpu implementation is nor specialized for p=2, can you please fix that?

**Comment by jjsjann123** [comment]
Let me see if cpu implementation actually benefits from specialized instruction. :)

**Comment by jjsjann123** [comment]
I don't see speedup via special-case norm-2 on cpu perf, but it's also not slower so I'll keep the update.

**Review by ngimel — APPROVED** [review]

File: aten/src/ATen/native/cpu/ReduceOpsKernel.cpp [hunk 1]
```diff
@@ -84,6 +84,15 @@ static void norm_kernel_tensor_iterator_impl(
         scalar_t(0)
       );
     });
+  } else if (val == 2) {
+    printf("specialized norm");
```
Comment (ngimel) [inline aten/src/ATen/native/cpu/ReduceOpsKernel.cpp]: Print statement!

**Comment by jjsjann123** [comment]
Should I be concerned about internal failure?

**Comment by ngimel** [comment]
No, and backward compatibility is also a result of the different breakage.

**Pull request closed without merging**